internal/mirror/database/mirror.go

```diff
@@ -0,0 +1,223 @@
+				($1, $2, $3, $4)
+			RETURNING id
+		`, m.IndexFile, m.ExportRoot, m.CloudStorageBucket, m.FilenameRoot)
+
+		if err := row.Scan(&m.ID); err != nil {
+			return fmt.Errorf("fetching mirror.ID: %w", err)
+		}
+		return nil
+	})
+}
+
+func (db *MirrorDB) DeleteMirror(ctx context.Context, m *model.Mirror) error {
+	return db.db.InTx(ctx, pgx.Serializable, func(tx pgx.Tx) error {
+		_, err := tx.Exec(ctx, `
+			DELETE FROM
+				MirrorFile
+			WHERE
+				mirror_id = $1
+			`, m.ID)
+		if err != nil {
+			return fmt.Errorf("failed to delete mirror files: %w", err)
+		}
+
+		_, err = tx.Exec(ctx, `
+			DELETE FROM
+				Mirror
+			WHERE
+				id = $1
+			`, m.ID)
+		if err != nil {
+			return fmt.Errorf("failed to delete mirror config: %w", err)
+		}
+
+		return nil
+	})
+}
+
+func (db *MirrorDB) Mirrors(ctx context.Context) ([]*model.Mirror, error) {
+	var mirrors []*model.Mirror
+
+	if err := db.db.InTx(ctx, pgx.ReadCommitted, func(tx pgx.Tx) error {
+		rows, err := tx.Query(ctx, `
+			SELECT
+				id, index_file, export_root, cloud_storage_bucket, filename_root
+			FROM
+				mirror
+		`)
+		if err != nil {
+			return fmt.Errorf("failed to list: %w", err)
+		}
+		defer rows.Close()
+
+		for rows.Next() {
+			if err := rows.Err(); err != nil {
+				return fmt.Errorf("faield to iterate: %w", err)
+			}
+
+			var m model.Mirror
+			if err := rows.Scan(&m.ID, &m.IndexFile, &m.ExportRoot, &m.CloudStorageBucket, &m.FilenameRoot); err != nil {
+				return fmt.Errorf("reading row: %w", err)
+			}
+			mirrors = append(mirrors, &m)
+		}
+		return nil
+	}); err != nil {
+		return nil, fmt.Errorf("listing mirror configs: %w", err)
+	}
+
+	return mirrors, nil
+}
+
+func (db *MirrorDB) SaveFiles(ctx context.Context, mirrorID int64, filenames []string) error {
+	const deleteName = "delete mirror file"
+	const insertName = "insert mirror file"
+
+	wantFiles := make(map[string]struct{}, len(filenames))
+	for _, fname := range filenames {
+		wantFiles[fname] = struct{}{}
+	}
+
+	return db.db.InTx(ctx, pgx.ReadCommitted, func(tx pgx.Tx) error {
+		knownFiles, err := readFiles(ctx, tx, mirrorID)
+		if err != nil {
+			return err
+		}
+
+		toDelete := make([]*model.MirrorFile, 0)
+
+		for _, mirrorFile := range knownFiles {
+			if _, ok := wantFiles[mirrorFile.Filename]; ok {
+				delete(wantFiles, mirrorFile.Filename)
+			} else {
+				toDelete = append(toDelete, mirrorFile)
+			}
+		}
+
+		if len(toDelete) > 0 {
+			if _, err := tx.Prepare(ctx, deleteName, `
+				DELETE FROM
+						MirrorFile
+				WHERE
+					mirror_id = $1 AND filename = $2
+				`); err != nil {
+				return fmt.Errorf("failed to prepare delete DB statement: %w", err)
+			}
+
+			for _, mf := range toDelete {
+				if _, err := tx.Exec(ctx, deleteName, mf.MirrorID, mf.Filename); err != nil {
+					return fmt.Errorf("failed to delete mirrofile: %w", err)
+				}
+			}
+		}
+
+		// Create files if they still need to be created.
+		if len(wantFiles) > 0 {
+			if _, err := tx.Prepare(ctx, insertName, `
```

Comment:
This is the same transaction as the delete, right? If wantFiles and toDelete include the same (mirror_id, filename), what happens?

Reply:
wantFiles only contains files that weren't read back into knownFiles. 

added comments to try and make this clearer